hibernate-reactive-core/src/main/java/org/hibernate/reactive/common/NativeQuery.java

```diff
@@ -0,0 +1,87 @@
+/* Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive.common;
+
+/**
+ * A description of an executable native SQL query, including the query
+ * string, a {@link ResultSetMapping} or return type, and a list of
+ * affected tables and/or entities which should be synchronized before
+ * execution of the query.
+ *
+ * @param <T> the return type of the query
+ *
+ * @author Gavin King
+ */
+public class NativeQuery<T> {
+    private static final Class<?>[] NO_ENTITIES = new Class[0];
+    private static final String[] NO_TABLES = new String[0];
+
+    private final String queryString;
+    private final ResultSetMapping<T> resultSetMapping;
+    private final String[] queryTables;
+    private final Class<?>[] queryEntities;
+    private final Class<T> returnType;
+
+    private NativeQuery(String queryString,
+                        ResultSetMapping<T> resultSetMapping,
+                        String[] queryTables,
+                        Class<?>[] queryEntities,
+                        Class<T> returnType) {
+        this.queryString = queryString;
+        this.resultSetMapping = resultSetMapping;
+        this.queryTables = queryTables;
+        this.queryEntities = queryEntities;
+        this.returnType = returnType;
+    }
+
+    public static <T> NativeQuery<T> create(String queryString, ResultSetMapping<T> resultSetMapping) {
+        return create(queryString, resultSetMapping, NO_TABLES);
+    }
+
+    public static <T> NativeQuery<T> create(String queryString, ResultSetMapping<T> resultSetMapping, String[] queryTables) {
```

Comment:
Now it might be more suitable to use `...` ?

It feels like you purposefully introduced the mandatory `ResultSetMapping` parameter (or `Class<T> returnType` below) to make it viable?

Reply:
Alternatively, I'd use a builder DSL.

Reply:
> Now it might be more suitable to use `...` ?

I agree, and I considered it, but I'm just wondering if it might get in the way of adding additional overloads in future. TBH I'm just not sure.

> Alternatively, I'd use a builder DSL.

I have two objections to that:

1. a kinda principled objection to using a mutable object when an immutable object will do the job, and
2. also the practical objection to how verbose that's going to get: it's already quite difficult to format the code for a native query call in HR with all the surrounding anonymous functions..

Reply:
I generally don't like DSL builder objects on the "hot path" from a performance perspective, but I'm thinking here that the expected usage is to define native queries as application constants.

So I could see myself liking the shape of:

	private static NativeQuery<Object[]> TITLES_AND_AUTHORS_QUERY = NativeQuery.fromSQL("select b.title, a.name from books b join authors a on author_id=a.id order by b.isbn")
		.withResultSetMapping(Object[].class) // Optional : Object[] as default return type for T of NativeQuery<T> ?
		.affectsTableSpaces("authors","books") // Optional
		.create();
 
But N.B. I'm somewhat cheating here, as I see the current API requires a Session to define & validate a `ResultSetMapping`. So maybe this requires a bit more work, but it has the merit of allowing us to push such validations at build time & also improve efficiency.

Ideally this could be assigned to a `Query<T>` rather than a `NativeQuery<T>` and could tie in with more advanced build time validations, or allow people to use a different builder.

> I agree, and I considered it, but I'm just wondering if it might get in the way of adding additional overloads in future. TBH I'm just not sure.

It's precisely to allow for future overloads that I'd go with a builder in this case.

Reply:
> I generally don't like DSL builder objects on the "hot path" from a performance perspective, but I'm thinking here that the expected usage is to define native queries as application constants.

Hrm. I don't hate it.

> But N.B. I'm somewhat cheating here, as I see the current API requires a `Session` to define & validate a `ResultSetMapping`

I'm not actually happy with this, FTR, I actually want to have a code generator create type safe references to things like named queries and `ResultSetMapping`s.

> It's precisely to allow for future overloads that I'd go with a builder in this case.

Yesyes, I understand that. I like constructors for immutability. On the other hand, it *is* possible to make immutable builders, so perhaps we could do it that way.

Reply:
Sounds like we're on the same page, but I'm not sure about the conclusion?

I think I'd go with the Builder DSL, not because it would make all such things possible now, but it would allow us to get there eventually.

Reply:
> I'm not sure about the conclusion?

Well I figured I would try it out and see what I think.